### ISSUE TITLE
CI Failure due to a quickcheck counter-example 

### ISSUE THREAD

**Issue by Kleidukos** [haskell/haddock#1142]
It would seem like dae8d9de found a counter-example in the CI, [effectively crashing the test suite](https://travis-ci.org/github/haskell/haddock/builds/664474529#L2006):

```
  test/Documentation/Haddock/ParserSpec.hs:484:7: 

  1) Documentation.Haddock.Parser.parseParas, when parsing text paragraphs, preserve all regular characters

       uncaught exception: ErrorCall

       Haddock.Parser.parse: "<haddock>" (line 1, column 1):

       unexpected '\8197'

       expecting "\n", ">>>", "@since ", "*", "-", "(", digit, ">", "@", "prop>", "[" or end of input

       string: Failed to match the input string

       CallStack (from HasCallStack):

         error, called at src/Documentation/Haddock/Parser.hs:103:11 in main:Documentation.Haddock.Parser

       (after 7 tests and 2 shrinks)

         "\8197"

  To rerun use: --match "/Documentation.Haddock.Parser/parseParas/when parsing text paragraphs/preserve all regular characters/"
```

The quickcheck seed is the following `552694385`

cc @harpocrates

**Comment by harpocrates** [comment]
Several issues that need to be fixed here:

  * `Documentation.Haddock.Parser.parseParas Nothing "\8197"` causes a crash! This function should never fail, so this is a bug! The root of the issue seems to be in `Documentation.Haddock.Parser.Util.horizontalSpace` which doesn't account for unicode spaces. I propose that we turn `horizontalSpace` into a function:

    ```haskell
    -- | Characters that count as horizontal space
    horizontalSpace :: Char -> Bool
    horizontalSpace '\n' = False
    horizontalSpace c = isSpace c
    ```

  * The `preserve all regular characters` test case should filter out _all_ whitespace characters. Again, we need to use the `isSpace` function to catch unicode spaces.

  * We should add a new quickcheck test case which throws any arbitrary unicode input at Haddock and just checks that we don't crash.

**Comment by Kleidukos** [comment]
Side-question: Do I need Cabal 3 to run this command? I get the following error in my shell:

```sh
unrecognized 'v2-test' option `--test-option=--seed=552694385'
```

**Comment by harpocrates** [comment]
Yes, I think `--test-option` only (finally) got added in Cabal 3.

There is probably another way to do this without Cabal 3, but it'll take some more tweaking. The  workaround used before was to `cabal v2-run` the testsuite directly, then pass arguments to that (`cabal v2-run haddock-library:spec -- --seed=552694385` I think) If you are still using the `v1-` commands, the following might work:

```
cabal v1-test haddock-library:spec \
  --show-details=direct \
  --test-option='--seed=552694385
```

**Comment by Kleidukos** [comment]
Unfortunately it failed as well on the solver part:

```sh
❯ cabal v1-test haddock-library:spec \
  --show-details=direct \
  --test-option='--seed=552694385'
Resolving dependencies...
Warning: solver failed to find a solution:
Could not resolve dependencies:
[__0] trying: haddock-2.23.0 (user goal)
[__1] next goal: base (dependency of haddock)
[__1] rejecting: base-4.12.0.0/installed-4.1... (conflict: haddock =>
base^>=4.13.0.0)
[__1] fail (backjumping, conflict set: base, haddock)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: haddock, base
Trying configure anyway.
Configuring haddock-2.23.0...
cabal: Encountered missing dependencies:
base >=4.13.0.0 && <4.14, haddock-test ==0.0.1
```

But I will try the `v2-run` workaround, thank you :)

**Comment by harpocrates** [comment]
Fixed in b0514cc5d53bb37424177d2ba986216a914f8b1c.
